README.md
# ai_readme_generator

<p align="center"><img src="https://socialify.git.ci/tomkat-cr/ai_readme_generator/image?description=0&amp;font=Inter&amp;language=1&amp;name=1&amp;owner=1&amp;pattern=Plus&amp;stargazers=0&amp;theme=Light" alt="project-image"></p>

AI Readme Generator reads a Github repository and suggest a README.md file based on that code using Langchain and OpenAI's GPT4.

Main functionalities of this project are listed below:

1. The script clones a given repo from its url or local path, and fetches file content and metadata.
2. The content and metadata fetched are then passed to GPT-4 model to generate the README suggestion.
3. It can filter out files based on their extensions.
4. The script can also work in debug mode. In debug mode, the script provides more detailed output logs.

## Usage

Please note, in order to use this script, you will need the `OPENAI_API_KEY`.

To use the script from command line, please use the following template and fill in the relevant details:

```bash
pipenv run python -m ai_readme_generator.main -h
```

```bash
usage: main.py [-h] [-u REPO_URL] [-b BRANCH] [-e FILE_EXT_FILTER] [-m {gpt-3.5-turbo,gpt-3.5-turbo-16k,gpt-4,gpt-4-32k}]
               [-t TEMPERATURE] [-D DEBUG]

AI README Generator

options:
  -h, --help            show this help message and exit
  -u REPO_URL, --repo_url REPO_URL
                        Git repository URL or local path
  -b BRANCH, --branch BRANCH
                        Repository branch name. Default: main
  -e FILE_EXT_FILTER, --file_ext_filter FILE_EXT_FILTER
                        Filter the files based on their extensions. For instance: 'py,md', 'js,jsx' or 'php'
  -m {gpt-3.5-turbo,gpt-3.5-turbo-16k,gpt-4,gpt-4-32k}, --model {gpt-3.5-turbo,gpt-3.5-turbo-16k,gpt-4,gpt-4-32k}
                        GPT model. Default: gpt-3.5-turbo-16k
  -t TEMPERATURE, --temperature TEMPERATURE
                        Temperature for the GPT model. Default: 0.7
  -D DEBUG, --debug DEBUG
                        Show debug information
```

- `REPO_URL`: repository URL for which you want to generate README. It is a mandatory parameter. This can be also a local path, so repo won't be cloned in the `/tmp` directory.

- `BRANCH`: It is the branch of the repo from which you want to generate README. This is an optional parameter, the `main` branch will be used by default if no branch is specified.

- `FILE_EXT_FILTER`: It is to filter the files based on their extensions. For instance: `py,md`, `js,jsx` or `php`. This is an optional parameter, all files will be used if no filter is given.

For instance,

```bash
pipenv run python -m ai_readme_generator.main -u https://github.com/bard/ai-assistant -b main -e py,md
```

## Dependencies

Following Python libraries are required:
- git (Repo)
- langchain.document_loaders (GitLoader)
- openai
- os, sys
- dotenv (load_dotenv)
- pprint

## Installation Steps

1. Clone the repo:

```
git clone https://github.com/tomkat-cr/ai_readme_generator
```

2. Change the directory:

```
cd ai_readme_generator
```

3. Start the virtual environment:

```
python3 -m venv venv
source venv/bin/activate
```

4. Install requirements:

```
pip install -r requirements.txt
```

5. Create the .env file:

```
cp .env-example .env
vi .env
```

6. Configure parameters:

```
OPENAI_API_KEY=sk-XXXXXX
DEBUG=0
```

7. Run it (it'll ask for git repo URL/local path and branch):

```
pipenv run python -m ai_readme_generator.main
```

8. or run it with repo URL/path and branch

```
pipenv run python -m ai_readme_generator.main -u https://github.com/tomkat-cr/ai_readme_generator -b main
```

## Note

Please be aware that script will not attempt to delete or modify any files or directories outside of its designated working area.

Enjoy exploring the project!

## Contributions

Feel free to submit pull requests create issues or spread the word.

## Built with

Technologies used in the project:

- python3
- openai
- langchain
- gitpython
- pipenv
- make

## License

This project is licensed under the GNU 2.0


CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and [Keep a Changelog](http://keepachangelog.com/).



## Unreleased
---

### New

### Changes

### Fixes

### Breaks


## 0.1.4 (2023-07-07)

### New
Embeddings processing (1st approach).


## 0.1.3 (2023-07-05)

### New
New model and temperature parameters. Defaults: gpt-3.5-turbo-16k and 0.7.
Test generation: "prompt_type" parameter with values: "readme" to handle README.md generation and "test" for pytest generation.
Command line parameters changed from positional to named with the argparse.
Converted from functions to class: AiReadmeGenerator.


## 0.1.2 (2023-07-04)

### New
Configurate model and temperature.
Named parameters.


## 0.1.1 (2023-07-03)

### New
File extensions filter.
Allows to specify the repo URL, branch and file extesions filter from the command line.
Ask user for branch (default: main) and file extensions filter if no command line parameters are passed.
Ask user for openai api key if the environment variable OPENAI_API_KEY has no value.
Recognize "https://" prefix in repo url to know if it's a local or remote repo.
Program start header to identify repo URL and branch.
Print file extensions filter and files to be included.

### Changes

Update README.md.


## 0.1.0 (2023-07-30)
---

### New
Initial development of the AI Readme Generator. Reads a Github repository and suggest a README.md file based on the repo's code, using OpenAI GPT4.


ai_readme_generator/main.py
import os

from dotenv import load_dotenv
from pprint import pprint

from git import Repo
import openai
from langchain.document_loaders import GitLoader

from langchain.chat_models import ChatOpenAI
from langchain.memory import ConversationBufferMemory
from langchain.chains import ConversationalRetrievalChain
from langchain.indexes import VectorstoreIndexCreator

from ai_readme_generator.init_parser import init_parser


class AiReadmeGenerator():
    def __init__(self):
        parser = init_parser()
        args = parser.parse_args()
        self.repo_url = args.repo_url
        self.branch = args.branch
        self.file_ext_filter = args.file_ext_filter
        self.model = args.model
        self.temperature = args.temperature
        self.debug = args.debug == '1'
        self.prompt_type = args.prompt_type
        self.processing_method = "model_input"
        # self.processing_method = "embeddings"

    def get_def_values(self):
        load_dotenv()

        try:
            if os.getenv("OPENAI_API_KEY"):
                self.openai_api_key = os.getenv("OPENAI_API_KEY")
            else:
                self.openai_api_key = str(input("Enter OPENAI_API_KEY: "))
        except KeyboardInterrupt:
            return "Keyboard interrupt"
        except Exception as err:
            return str(err)

        if not self.repo_url:
            try:
                self.repo_url = str(input(
                    "Git repository URL: "
                ))
                self.branch = str(input(
                    "Branch (default 'main'): "
                ))
                self.file_ext_filter = str(input(
                    "file extension filter (default all files): "
                ))
                self.model = str(input(
                    "GPT model (gpt-3.5-turbo, gpt-3.5-turbo-16k," +
                    " gpt-4 or gpt-4-32k): "
                ))
                self.temperature = str(input(
                    "Temperature for the GPT model (default: 0.7): "
                ))
                self.prompt_type = str(input(
                    "Prompt type (readme|test): "
                ))
            except KeyboardInterrupt:
                return "Keyboard interrupt"
            except Exception as err:
                return str(err)

        if not self.temperature:
            self.temperature = 0.7
        else:
            self.temperature = float(self.temperature)
        if not self.model:
            self.model = "gpt-3.5-turbo-16k"
        if not self.prompt_type:
            self.prompt_type = "readme"
        elif self.prompt_type not in ["readme", "test"]:
            return "ERROR: Invalid Prompt type."

        return ""

    def remove_dir(self, local_temp_path):
        if local_temp_path == '/':
            raise Exception("Cannot use / as a temp path")
        if not os.path.exists(local_temp_path):
            return
        if os.path.isfile(local_temp_path):
            os.remove(local_temp_path)
            return
        for root, dirs, files in os.walk(local_temp_path, topdown=False):
            for name in files:
                os.remove(os.path.join(root, name))
            for name in dirs:
                os.rmdir(os.path.join(root, name))

    def get_repo(self):
        response = {
            "error": False,
            "error_msg": "",
        }
        if self.repo_url[:8] == "https://":
            # https://python.langchain.com/docs/modules/data_connection/document_loaders/integrations/git
            local_temp_repo_path = f"/tmp/{self.repo_url.rsplit('/', 1)[1]}"
            try:
                self.remove_dir(local_temp_repo_path)
            except Exception as err:
                if self.debug:
                    raise
                response["error"] = True
                response["error_msg"] = f"ERROR: {str(err)}"
                return response
            try:
                repo = Repo.clone_from(
                    url=self.repo_url,
                    to_path=local_temp_repo_path
                )
            except Exception as err:
                if self.debug:
                    raise
                response["error"] = True
                response["error_msg"] = f"ERROR: {str(err)}"
                return response
        else:
            try:
                local_temp_repo_path = self.repo_url
                repo = Repo(local_temp_repo_path)
            except Exception as err:
                if self.debug:
                    raise
                response["error"] = True
                response["error_msg"] = f"ERROR: {str(err)}"
                return response
        if not self.branch:
            try:
                self.branch = repo.head.reference
            except Exception as err:
                if self.debug:
                    raise
                response["error"] = True
                response["error_msg"] = f"ERROR: {str(err)}"
                return response
        print()
        print(f"Github repository URL: {self.repo_url}")
        print(f"Branch (default 'main'): {self.branch}")
        try:
            loader = GitLoader(
                repo_path=f"{local_temp_repo_path}/",
                branch=self.branch
            )
        except Exception as err:
            if self.debug:
                raise
            response["error"] = True
            response["error_msg"] = f"ERROR: {str(err)}"
            return response
        response["data"] = loader.load()
        if self.debug:
            print(f"Repo: {self.repo_url}")
            print(f"Branch: {self.branch}")
            print("Data:")
            pprint(response["data"])
        return response

    def get_prompt(self, pcode):
        if pcode == "readme":
            prompt = \
                """Write a readme.md file for this repository content.
                """
        if pcode == "test":
            prompt = \
                """Give me a pytest file for this repository content.
                """
        return prompt

    def std_response(self):
        response = {
            "error": False,
            "content": "",
        }
        return response

    def model_input_method(self):
        response = self.std_response()
        text = ""
        for doc_obj in self.repo_data:
            if self.file_ext_filter and \
               not doc_obj.metadata["file_type"] \
               in self.file_extensions_allowed:
                continue
            text += doc_obj.page_content
            text += str(doc_obj.metadata)
            if self.debug:
                print(doc_obj.metadata["source"])

        # https://platform.openai.com/docs/api-reference/completions/create
        # https://github.com/openai/openai-cookbook/blob/main/examples/Question_answering_using_embeddings.ipynb

        openai.api_key = self.openai_api_key
        prompt = self.get_prompt(self.prompt_type)
        messages = [
            {
                "role": "system",
                "content": prompt
            },
            {"role": "user", "content": text},
        ]

        print("Model Input processing start...")

        try:
            ai_response = openai.ChatCompletion.create(
                model=self.model,
                temperature=self.temperature,
                messages=messages,
            )
        except Exception as err:
            if self.debug:
                raise
            response["error"] = True
            response["content"] = f"ERROR: {str(err)}"

        print("Model Input processing done!")

        if not response["error"]:
            try:
                response["content"] = \
                    ai_response["choices"][0]["message"]["content"]
            except Exception as err:
                response["error"] = True
                response["content"] = f"ERROR: {str(err)}"

        return response

    def embeddings_method(self):
        response = self.std_response()
        prompt = self.get_prompt(self.prompt_type)

        print("Embedding processing start...")

        index = VectorstoreIndexCreator().from_documents(
            self.repo_data
        )
        llm = ChatOpenAI(
            model=self.model,
            temperature=self.temperature,
        )
        memory = ConversationBufferMemory(
            memory_key='chat_history', return_messages=True
        )
        conversation_chain = ConversationalRetrievalChain.from_llm(
            llm=llm,
            retriever=index.vectorstore.as_retriever(),
            memory=memory
        )

        print("LLM processing...")

        ai_response = conversation_chain({'question': prompt})
        chat_history = ai_response['chat_history']

        response["content"] = ""
        for i, message in enumerate(chat_history):
            if i % 2 == 0:
                continue
            else:
                response["content"] += message.content

        print("Embedding processing done!")

        return response

    def get_readme_suggestion(self):
        """Gets a readme.md file suggestion from the given
           GitHub repository URL or local path.
        """
        response = {
            "error": False,
            "content": "",
        }
        repo_response = self.get_repo()
        if repo_response["error"]:
            response["error"] = True
            response["content"] = repo_response["error_msg"]
            return response
        self.repo_data = repo_response["data"]

        self.file_extensions_allowed = None
        if self.file_ext_filter:
            self.file_extensions_allowed = \
                [f".{v}" for v in self.file_ext_filter.split(",")]
            print(f"File extensions allowed: {self.file_extensions_allowed}")
            if self.debug:
                print()
                print("Files to be included in the context:")

        if self.processing_method == "embeddings":
            response = self.embeddings_method()
        else:
            response = self.model_input_method()

        if self.debug:
            print("The response is:")
            pprint(response)
        return response

    def main(self):
        print()
        print("AI README.md file generator")
        error_msg = self.get_def_values()
        if error_msg:
            print()
            print(error_msg)
            return
        readme_suggestion = self.get_readme_suggestion()
        print()
        if readme_suggestion["error"]:
            print("Could not generate the suggested README.md:")
        else:
            print(f"The suggested {self.prompt_type} is:")
        print()
        print(readme_suggestion["content"])


if __name__ == "__main__":
    airg = AiReadmeGenerator()
    airg.main()


scripts/install_dev_tools.sh
#!/bin/bash
ERROR_MSG=""

if [ "${ERROR_MSG}" = "" ]; then
    export SCRIPTS_DIR="$(dirname "$0")";
    cd "${SCRIPTS_DIR}"
    export SCRIPTS_DIR="$(pwd)";
    cd "${SCRIPTS_DIR}/.."
    export BASE_DIR="$(pwd)";
    if [ -f "${BASE_DIR}/.env" ]; then
        set -o allexport; . "${BASE_DIR}/.env" ; set +o allexport ;
    fi
fi

if [ "${PYTHON_VERSION}" == "" ]; then
    PYTHON_VERSION="3.9.17"
fi

BASH_PROFILE="${HOME}/.bash_profile"
BASH_SOURCE1="bash"
if [ -f "${HOME}/.zshrc" ]; then
    BASH_PROFILE="${HOME}/.zshrc"
    BASH_SOURCE1="zsh"
fi

PYENV_INSTALL_PREFIX=""
MACHINE_ARCHITECTURE=$(arch)
if [ "${MACHINE_ARCHITECTURE}" = "arm64" ]; then
    PYENV_INSTALL_PREFIX="arch -x86_64"
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! pyenv --version
    then
        # install pyenv
        echo ""
        echo "Install pyenv"
        echo ""
        curl https://pyenv.run | bash
        if ! pyenv --version
        then
            ERROR_MSG="ERROR: pyenv could not be installed."
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! grep -q 'eval "$(pyenv init --path)"' ${BASH_PROFILE}; then
        # add pyenv to the PATH and enable shims and autocompletion
        echo ""
        echo "Add pyenv to the PATH and enable shims and autocompletion"
        echo ""

        echo "" >> ${BASH_PROFILE}
        echo "# `date`" >> ${BASH_PROFILE}
        echo "# Add pyenv to the PATH and enable shims and autocompletion" >> ${BASH_PROFILE}
        echo "" >> ${BASH_PROFILE}
        echo 'export PYENV_ROOT="${HOME}/.pyenv"' >> ${BASH_PROFILE}
        echo 'export PATH="$PYENV_ROOT/bin:$PATH"' >> ${BASH_PROFILE}
        echo 'eval "$(pyenv init --path)"' >> ${BASH_PROFILE}
        echo 'eval "$(pyenv init -)"' >> ${BASH_PROFILE}
        # echo 'eval "$(pyenv virtualenv-init -)"' >> ${BASH_PROFILE}
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if [ ! -d "${HOME}/.pyenv/versions/${PYTHON_VERSION}" ]; then
        # install python version using pyenv
        echo ""
        echo "Install python with: pyenv install ${PYTHON_VERSION}"
        echo ""
        if ! ${PYENV_INSTALL_PREFIX} pyenv install ${PYTHON_VERSION}
        then
            ERROR_MSG="ERROR: running pyenv install ${PYTHON_VERSION}"
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    echo ""
    echo "Changing python version: pyenv local ${PYTHON_VERSION}"
    echo ""
    if ! pyenv local ${PYTHON_VERSION}
    then
        ERROR_MSG="ERROR: running pyenv local ${PYTHON_VERSION}"
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! pipenv --version
    then
        # install pipenv
        echo ""
        echo "Install pipenv"
        echo ""
        if ! pip install --user pipenv
        then
            ERROR_MSG="ERROR: pipenv could not be installed."
        else
            # add pipenv to the bash profile
            ADD_CMD_TO_PROFILE="eval \"$(_PIPENV_COMPLETE=${BASH_SOURCE1}_source pipenv)\""
            if ! grep -q "${ADD_CMD_TO_PROFILE}" ${BASH_PROFILE}; then
                echo ""
                echo "Add pipenv to the ${BASH_SOURCE1} profile"
                echo ""

                echo "" >> ${BASH_PROFILE}
                echo "# `date`" >> ${BASH_PROFILE}
                echo "# Add pipenv init" >> ${BASH_PROFILE}
                echo "" >> ${BASH_PROFILE}
                echo "${ADD_CMD_TO_PROFILE}" >> ${BASH_PROFILE}
            fi
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! make --version
    then
        # install make
        echo ""
        echo "Install make"
        echo ""

        if [[ "$OSTYPE" == "darwin"* ]]; then
            brew install make
        else
            if ! apt -y install make
            then
                if ! yum -y install make
                then
                    if ! apk add make
                    then
                        ERROR_MSG="ERROR: make could not be installed in linux (debian/rhel/alpine)"
                    fi
                fi
            fi  
        fi
        if ! make --version
        then
            ERROR_MSG="ERROR: make could not be installed"
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! aws --version
    then
        sh ${SCRIPTS_DIR}/install_aws_cli.sh
        if ! aws --version
        then
            ERROR_MSG="ERROR: aws-cli could not be installed"
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    echo ""
    echo "Reload the shell"
    echo ""
    exec "$SHELL"
fi

echo ""
if [ "${ERROR_MSG}" = "" ]; then
    echo "Done!"
else
    echo "${ERROR_MSG}" ;
fi
echo ""


ai_readme_generator/init_parser.py
import argparse


def init_parser():
    parser = argparse.ArgumentParser(
        description='AI README Generator',
    )
    parser.add_argument(
        "-u", "--repo_url",
        default='',
        required=False,
        help="Git repository URL or local path",
    )
    parser.add_argument(
        "-b", "--branch",
        default='',
        required=False,
        help="Repository branch name. Default: main",
    )
    parser.add_argument(
        "-e", "--file_ext_filter",
        default='',
        required=False,
        help="Filter the files based on their extensions." +
             " For instance: 'py,md', 'js,jsx' or 'php'",
    )
    parser.add_argument(
        "-m", "--model",
        default='gpt-3.5-turbo-16k',
        required=False,
        choices=[
            "gpt-3.5-turbo", "gpt-3.5-turbo-16k", "gpt-4", "gpt-4-32k"
        ],
        help="GPT model. Default: gpt-3.5-turbo-16k",
    )
    parser.add_argument(
        "-t", "--temperature",
        default='0.7',
        required=False,
        help="Temperature for the GPT model. Default: 0.7",
    )
    parser.add_argument(
        "-pt", "--prompt_type",
        default='readme',
        choices=["readme", "test"],
        required=False,
        help="Prompt type: 'readme' to generate the README.md file" +
             ", 'test' to generate the testing code based on pytest.",
    )
    parser.add_argument(
        "-D", "--debug",
        default='0',
        required=False,
        help="Show debug information",
    )

    return parser
